src/NServiceBus.RabbitMQ/Receiving/MessagePump.cs

```diff
@@ -140,107 +139,77 @@ void Connection_ConnectionShutdown(object sender, ShutdownEventArgs e)
                 return;
             }
 
+            try
+            {
+                await ProcessMessage(eventArgs).ConfigureAwait(false);
+            }
+            finally
+            {
+                semaphore.Release();
+            }
+        }
+
+        private async Task ProcessMessage(BasicDeliverEventArgs message)
+        {
             Dictionary<string, string> headers = null;
             string messageId = null;
-            var pushMessage = false;
-
             try
             {
                 try
                 {
                     messageId = messageConverter.RetrieveMessageId(message);
                     headers = messageConverter.RetrieveHeaders(message);
-
-                    pushMessage = true;
                 }
                 catch (Exception ex)
                 {
                     await poisonMessageForwarder.ForwardPoisonMessageToErrorQueue(message, ex, settings.ErrorQueue).ConfigureAwait(false);
+                    await TryAcknowledgeMessage(message.DeliveryTag, messageId).ConfigureAwait(false);
+                    return;
                 }
 
-                var rejectMessage = false;
-
-                if (pushMessage)
-                {
-                    using (var tokenSource = new CancellationTokenSource())
-                    {
-                        var pushContext = new PushContext(messageId, headers, new MemoryStream(message.Body ?? new byte[0]), new TransportTransaction(), tokenSource, new ContextBag());
-
-                        await pipe(pushContext).ConfigureAwait(false);
-
-                        rejectMessage = tokenSource.IsCancellationRequested;
-                    }
-                }
-
-                if (rejectMessage)
-                {
-                    try
-                    {
-                        await RejectMessage(channel, message.DeliveryTag).ConfigureAwait(false);
-                    }
-                    catch (AlreadyClosedException ex)
-                    {
-                        Logger.Warn($"Attempt to reject message {messageId} failed because the channel was closed. The message will be requeued.", ex);
-                    }
-                }
-                else
+                using (var tokenSource = new CancellationTokenSource())
                 {
-                    try
+                    var pushContext = new PushContext(messageId, headers, new MemoryStream(message.Body ?? new byte[0]), new TransportTransaction(), tokenSource, new ContextBag());
+                    await pipe(pushContext).ConfigureAwait(false);
+                    if (tokenSource.IsCancellationRequested)
```

Comment:
Formatting: I'd prefer to have an empty line between the `pipe` call and the `if`.